crates/lang_srv/src/server.rs

```diff
@@ -61,34 +68,123 @@ impl RocLs {
                 range: None,
                 full: Some(SemanticTokensFullOptions::Bool(true)),
             });
-
+        let completion_provider = CompletionOptions {
+            resolve_provider: Some(false),
+            trigger_characters: Some(vec![".".to_string()]),
+            //TODO: what is this?
+            all_commit_characters: None,
+            work_done_progress_options: WorkDoneProgressOptions {
+                work_done_progress: None,
+            },
+        };
         ServerCapabilities {
             text_document_sync: Some(text_document_sync),
             hover_provider: Some(hover_provider),
             definition_provider: Some(OneOf::Right(definition_provider)),
             document_formatting_provider: Some(OneOf::Right(document_formatting_provider)),
             semantic_tokens_provider: Some(semantic_tokens_provider),
+            completion_provider: Some(completion_provider),
             ..ServerCapabilities::default()
         }
     }
 
     /// Records a document content change.
     async fn change(&self, fi: Url, text: String, version: i32) {
-        self.registry()
-            .apply_change(DocumentChange::Modified(fi.clone(), text));
+        let updating_result = self.inner.change(&fi, text, version).await;
 
-        let diagnostics = match std::panic::catch_unwind(|| self.registry().diagnostics(&fi)) {
-            Ok(ds) => ds,
-            Err(_) => return,
-        };
+        //The analysis task can be cancelled by another change coming in which will update the watched variable
+        if let Err(e) = updating_result {
+            debug!("cancelled change. Reason:{:?}", e);
+            return;
+        }
+        debug!("applied_change getting and returning diagnostics");
+
+        let diagnostics = self.inner.registry.diagnostics(&fi).await;
 
         self.client
             .publish_diagnostics(fi, diagnostics, Some(version))
             .await;
     }
+}
+
+impl Inner {
+    pub fn new() -> Inner {
+        Self {
+            registry: Registry::default(),
+        }
+    }
+
+    async fn registry(&self) -> &Registry {
+        &self.registry
+    }
+
+    async fn close(&self, _fi: Url) {
+        ()
+    }
+
+    pub async fn change(
+        &self,
+        fi: &Url,
+        text: String,
+        version: i32,
+    ) -> std::result::Result<(), String> {
+        debug!("V{:?}:starting change", version);
```

Comment:
This is super nice. A lot better than the mutual lock version. Thanks for doing this.

Reply:
Thankyou! Your prompting helped a lot. I definitely think this is much better than what i initially had here.